# --- T2-COPYRIGHT-NOTE-BEGIN ---
# This copyright note is auto-generated by ./scripts/Create-CopyPatch.
# 
# T2 SDE: package/.../gq/newer_openldap.patch
# Copyright (C) 2004 - 2005 The T2 SDE Project
# 
# More information can be found in the files COPYING and README.
# 
# This patch file is dual-licensed. It is available under the license the
# patched project is licensed under, as long as it is an OpenSource license
# as defined at http://www.opensource.org/ (e.g. BSD, X11) or under the terms
# of the GNU General Public License as published by the Free Software
# Foundation; either version 2 of the License, or (at your option) any later
# version.
# --- T2-COPYRIGHT-NOTE-END ---
I've made a patch to fix gq crashing on openldap 2.2 in 
util.c:gq_ldap_explode_dn
 
This should also take care of possible crashes in openldap 2.1 -- the code 
there was just broken. I hope things are self-explanatory.
 
Cheers, Kuba

From Kuba Ober,
http://sourceforge.net/mailarchive/message.php?msg_id=11069271

--- ./src/util.c.orig	2005-03-10 17:24:17.489874232 -0300
+++ ./src/util.c	2005-03-10 17:29:54.772599464 -0300
@@ -1902,12 +1902,19 @@
 /* OpenLDAP 2.1 both deprecated and broke ldap_explode_dn in one go (I
    won't comment on this).
 
+   kuba@ma...: I don"t know what"s broken, but ldap_explode_dn,
+   albeit deprecated, just works. At least for API versions 2004 and 3001,
+   i.e. OpenLDAP 2.1.30 and 2.2.13.
+
    NOTE: this is just a first try to adapt code from Pierangelo
    Masarati <masarati@aero.polimi.it>. 
 */
 
 char **gq_ldap_explode_dn(const char *dn, int dummy)
 {
+     #if (LDAP_API_VERSION >= 2004)
+     return ldap_explode_dn(dn, 0);
+     #else
      int i, rc;
      LDAPDN *parts;
      char **v = 0; 
@@ -1917,14 +1924,17 @@
      if (rc != LDAP_SUCCESS || parts == NULL) return NULL;
 
      /* count DN elements */
-     for( i = 0 ; parts[i] ; i++ ) ;
-     v = (char **) calloc((i + 2), sizeof(char*));
+     for( i = 0 ; parts[0][i] ; i++ ) ;     
+     v = (char **) malloc((i + 1) * sizeof(char*));
 
-     for( i = 0 ; parts[i] ; i++ ) {
+     for( i = 0 ; parts[0][i] ; i++ ) {
 	  ldap_rdn2str(parts[0][i], &v[i],
 		       LDAP_DN_FORMAT_LDAPV3 | LDAP_DN_PRETTY );
      } 
+     v[i] = 0;
+     ldap_dnfree(parts);
      return v;
+     #endif
 }
      
 void gq_exploded_free(char **exploded_dn)
